src/eigsolve/lanczos.jl

```diff
@@ -150,3 +150,141 @@ function eigsolve(A, x₀, howmany::Int, which::Selector, alg::Lanczos;
            vectors,
            ConvergenceInfo(converged, residuals, normresiduals, numiter, numops)
 end
+
+function eigsolve(A, x₀, howmany::Int, which::Selector, alg::BlockLanczos)
+    maxiter = alg.maxiter
+    krylovdim = alg.krylovdim
+    if howmany > krylovdim
+        error("krylov dimension $(krylovdim) too small to compute $howmany eigenvalues")
+    end
+    tol = alg.tol
+    verbosity = alg.verbosity
+
+    # The initial block size is determined by the number of the starting vectors provided by the user
+    S = typeof(inner(x₀[1], x₀[1])) # Scalar type of the inner product between vectors in x₀
+    block0 = BlockVec{S}(x₀)
+    bs = length(block0)
+    bs < 1 && error("The length of the starting vector x₀ must be greater than 0")
+
+    iter = BlockLanczosIterator(A, block0, krylovdim + bs, alg.qr_tol, alg.orth)
+    fact = initialize(iter; verbosity=verbosity)  # Returns a BlockLanczosFactorization
+    numops = 1    # Number of matrix-vector multiplications (for logging)
+    numiter = 1
+
+    converged = 0
+    local normresiduals, D, U
+
+    while true
+        K = length(fact)
+        β = normres(fact)
+
+        if β < tol && K < howmany && verbosity >= WARN_LEVEL
+            msg = "Invariant subspace of dimension $(K) (up to requested tolerance `tol = $tol`), "
+            msg *= "which is smaller than the number of requested eigenvalues (i.e. `howmany == $howmany`)."
+            @warn msg
+        end
+        # BlockLanczos can access the case of K = 1 and doesn't need extra processing
+        if K >= krylovdim || β <= tol || (alg.eager && K >= howmany)
+            # Compute eigenvalues
+            # Note: Fast eigen solver for block tridiagonal matrices is not implemented yet.
+            BTD = view(fact.T, 1:K, 1:K)
+            D, U = eigen(Hermitian(BTD))
+            by, rev = eigsort(which)
+            p = sortperm(D; by=by, rev=rev)
+            D, U = permuteeig!(D, U, p)
+
+            # Detect convergence by computing the residuals
+            bs_r = fact.r_size   # The block size of the residual (decreases as the iteration goes)
+            r = residual(fact)
+            UU = U[(end - bs_r + 1):end, :]  # The last bs_r rows of U, used to compute the residuals
+            normresiduals = let R = block_inner(r, r)
+                map(u -> sqrt(real(dot(u, R, u))), cols(UU))
+            end
+            converged = count(<=(tol), normresiduals)
```

Comment:
In principle, it could happen that e.g. the first and third eigenvector are "converged" (i.e. have normresidual < tol), but not the second. If `howmany==2`, then this would trigger the termination of the algorithm, but return the first two eigenvalues and eigenvectors, where the second one has not converged. So probably
`converged = findfirst(>(tol), normresiduals) - 1` is better. But `findfirst` can return nothing, if all `normresiduals` would be `<= tol`. So this yields (untested)

```suggestion
            converged = something(findfirst(>(tol), normresiduals), K+1) - 1
```

Reply:
Alternatively, you can write an explicit loop as is done in `Lanczos`:
```suggestion
                converged = 0
                while converged < K && normresiduals[converged + 1] <= tol
                    converged += 1
                end
```

Reply:
You're right! When `(D[1], D[howmany])` contains `0` this can indeed happen.